File: core/document-engine/branding.py
"""Load society branding from site-meta.json for compose shells and stationery defaults."""

from __future__ import annotations

import html
import json
import pathlib
from typing import Any


def _cache_bust(path: str, meta: dict[str, Any]) -> str:
    if not path:
        return path
    if "?v=" in path or "&v=" in path:
        return path
    stamp = meta.get("lastUpdated") or meta.get("version") or ""
    if not stamp:
        return path
    sep = "&" if "?" in path else "?"
    return f"{path}{sep}v={stamp}"


def load_society_branding(site_root: pathlib.Path | str) -> dict[str, Any]:
    """Return normalized branding for compose simple shell and stationery defaults."""
    root = pathlib.Path(site_root).resolve()
    meta_path = root / "site-meta.json"
    meta: dict[str, Any] = {}
    if meta_path.is_file():
        try:
            parsed = json.loads(meta_path.read_text(encoding="utf-8"))
            if isinstance(parsed, dict):
                meta = parsed
        except (OSError, json.JSONDecodeError):
            meta = {}

    compose = meta.get("composeBranding") if isinstance(meta.get("composeBranding"), dict) else {}
    society = str(meta.get("societyName") or compose.get("societyName") or "").strip()
    colony = str(
        meta.get("siteName")
        or meta.get("brandName")
        or compose.get("colonyName")
        or ""
    ).strip()
    origin = str(meta.get("publicOrigin") or compose.get("publicOrigin") or "").strip().rstrip("/")
    email = str(compose.get("email") or meta.get("email") or "").strip()
    address = str(compose.get("addressLine") or compose.get("address") or "").strip()
    footer = str(compose.get("footerLine") or compose.get("footer") or "").strip()
    if not footer and society:
        footer = society

    logo_print = _cache_bust(
        str(meta.get("logoPrint") or compose.get("logoPrint") or "/assets/favicon-192.png"),
        meta,
    )
    logo_wm = _cache_bust(
        str(meta.get("logoWatermark") or compose.get("logoWatermark") or logo_print),
        meta,
    )

    meta_bits: list[str] = []
    if address:
        meta_bits.append(address)
    if email:
        meta_bits.append(email)
    if origin:
        host = origin.replace("https://", "").replace("http://", "")
        if host and host not in " · ".join(meta_bits):
            meta_bits.append(host)
    meta_line = " · ".join(meta_bits)

    return {
        "societyName": society or colony or "Residents Welfare Association",
        "colonyName": colony or society or "Society",
        "addressLine": address,
        "email": email,
        "publicOrigin": origin,
        "metaLine": meta_line,
        "footerLine": footer,
        "logoPrint": logo_print if logo_print.startswith("/") else f"/{logo_print.lstrip('/')}",
        "logoWatermark": logo_wm if logo_wm.startswith("/") else f"/{logo_wm.lstrip('/')}",
    }


def simple_compose_shell_html(branding: dict[str, Any], body_html: str, page_margin: str) -> tuple[str, str, str]:
    """Return (header_inner_html, footer_html, logo_src) for the simple compose chrome."""
    society = branding.get("societyName") or "Society"
    colony = branding.get("colonyName") or society
    meta_line = branding.get("metaLine") or ""
    footer = branding.get("footerLine") or society
    logo = branding.get("logoPrint") or "/assets/favicon-192.png"
    header = (
        f"<header class=\"org\">"
        f"<img src=\"{html.escape(str(logo))}\" alt=\"\">"
        f"<h1>{html.escape(str(society))}</h1>"
        f"<p class=\"sub\">{html.escape(str(colony))}</p>"
    )
    if meta_line:
        header += f"<p class=\"meta\">{html.escape(str(meta_line))}</p>"
    header += "</header>"
    foot = f"<footer class=\"foot\">{html.escape(str(footer))}</footer>"
    return header, foot, logo

File: sites/hbcsanyard/composer/branding.js
/**
 * Society branding from site-meta.json — drives simple chrome and stationery defaults.
 */

let siteBranding = null;

function cacheBust(path, meta) {
  if (!path) return path;
  if (/[?&]v=/.test(path)) return path;
  const stamp = meta?.lastUpdated || meta?.version || '';
  if (!stamp) return path;
  const sep = path.includes('?') ? '&' : '?';
  return `${path}${sep}v=${encodeURIComponent(String(stamp))}`;
}

function normalizeBranding(meta) {
  const m = meta && typeof meta === 'object' ? meta : {};
  const compose = m.composeBranding && typeof m.composeBranding === 'object' ? m.composeBranding : {};
  const society = String(m.societyName || compose.societyName || '').trim();
  const colony = String(m.siteName || m.brandName || compose.colonyName || '').trim();
  const origin = String(m.publicOrigin || compose.publicOrigin || '').trim().replace(/\/$/, '');
  const email = String(compose.email || m.email || '').trim();
  const address = String(compose.addressLine || compose.address || '').trim();
  let footer = String(compose.footerLine || compose.footer || '').trim();
  if (!footer && society) footer = society;
  const logoPrint = cacheBust(String(m.logoPrint || compose.logoPrint || '/assets/favicon-192.png'), m);
  const logoWatermark = cacheBust(String(m.logoWatermark || compose.logoWatermark || logoPrint), m);
  const metaBits = [];
  if (address) metaBits.push(address);
  if (email) metaBits.push(email);
  if (origin) {
    const host = origin.replace(/^https?:\/\//, '');
    if (host && !metaBits.join(' · ').includes(host)) metaBits.push(host);
  }
  return {
    societyName: society || colony || 'Residents Welfare Association',
    colonyName: colony || society || 'Society',
    addressLine: address,
    email,
    publicOrigin: origin,
    metaLine: metaBits.join(' · '),
    footerLine: footer,
    logoPrint: logoPrint.startsWith('/') ? logoPrint : `/${logoPrint.replace(/^\//, '')}`,
    logoWatermark: logoWatermark.startsWith('/') ? logoWatermark : `/${logoWatermark.replace(/^\//, '')}`,
  };
}

export function setSocietyBranding(meta) {
  siteBranding = normalizeBranding(meta);
}

export function getSocietyBranding() {
  return siteBranding || normalizeBranding(null);
}

export function simpleChromeHeadHtml() {
  const b = getSocietyBranding();
  const meta = b.metaLine ? `<p class="meta">${escapeHtml(b.metaLine)}</p>` : '';
  return `<div class="mhws-simple-chrome">
  <div class="mhws-simple-head">
  <img src="${escapeHtml(b.logoPrint)}" alt="">
  <h1>${escapeHtml(b.societyName)}</h1>
  <p class="sub">${escapeHtml(b.colonyName)}</p>
  ${meta}
</div>
<div class="rule" aria-hidden="true"><span class="pip"></span></div>
</div>`;
}

export function simpleChromeFootHtml() {
  const b = getSocietyBranding();
  return `<div class="mhws-simple-foot">${escapeHtml(b.footerLine)}</div>`;
}

export function defaultStationeryFromBranding() {
  const b = getSocietyBranding();
  return {
    paper: { id: 'A4', widthMm: 210, heightMm: 297, orientation: 'portrait' },
    backgroundColor: '#ffffff',
    logo: { src: b.logoPrint, widthMm: 18, align: 'center', enabled: true },
    headerLines: [
      {
        text: b.societyName,
        font: 'garamond',
        sizePt: 15,
        weight: '700',
        color: '#0b2a56',
        align: 'center',
      },
      {
        text: b.colonyName,
        font: 'source',
        sizePt: 10,
        weight: '600',
        color: '#1a6b3a',
        align: 'center',
      },
      ...(b.addressLine ? [{
        text: b.addressLine,
        font: 'noto',
        sizePt: 8,
        weight: '400',
        color: '#5a6a80',
        align: 'center',
      }] : []),
    ],
    watermark: { src: b.logoWatermark, opacity: 0.72, enabled: true },
    footer: {
      text: b.footerLine,
      font: 'noto',
      sizePt: 8,
      weight: '600',
      color: '#5a6a80',
    },
    border: { style: 'tricolor' },
  };
}

function escapeHtml(value) {
  return String(value || '')
    .replace(/&/g, '&amp;')
    .replace(/</g, '&lt;')
    .replace(/>/g, '&gt;')
    .replace(/"/g, '&quot;');
}
